File: src/scanner.js
const axios = require('axios');
const chalk = require('chalk');

class PrototypePollutionScanner {
  constructor(targetUrl, options = {}) {
    this.targetUrl = targetUrl;
    this.options = {
      method: 'POST', // Default to POST as it's more common for prototype pollution
      headers: {
        'Content-Type': 'application/json'
      },
      timeout: 5000,
      verbose: false,
      ...options
    };
    this.vulnerable = false;
    this.testPayloads = this.generateTestPayloads();
  }

  generateTestPayloads() {
    // Common prototype pollution test payloads
    return [
      { '__proto__': { 'polluted': true } },
      { 'constructor': { 'prototype': { 'polluted': true } } },
      { 'constructor': { 'prototype': { 'toString': 'polluted' } } },
      { 'prototype': { 'polluted': true } }
    ];
  }

  async scan() {
    console.log(chalk.blue(`[+] Starting prototype pollution scan for ${this.targetUrl}`));
    
    for (const payload of this.testPayloads) {
      try {
        if (this.options.verbose) {
          console.log(chalk.gray(`[.] Testing payload: ${JSON.stringify(payload)}`));
        }

        const response = await axios({
          method: this.options.method,
          url: this.targetUrl,
          data: payload,
          headers: this.options.headers,
          timeout: this.options.timeout
        });

        // Check if the pollution was successful by requesting a fresh object
        const checkResponse = await axios.get(this.targetUrl);
        
        if (this.checkForPollution(checkResponse.data)) {
          this.vulnerable = true;
          console.log(chalk.red(`[!] Prototype pollution vulnerability found with payload: ${JSON.stringify(payload)}`));
          console.log(chalk.red(`[!] Affected property: ${this.getPollutedProperty(checkResponse.data)}`));
          return true;
        }
      } catch (error) {
        if (this.options.verbose) {
          console.log(chalk.yellow(`[!] Error testing payload ${JSON.stringify(payload)}: ${error.message}`));
        }
      }
    }

    if (!this.vulnerable) {
      console.log(chalk.green(`[+] No prototype pollution vulnerabilities detected`));
    }
    return false;
  }

  checkForPollution(data) {
    try {
      // Try to create a new object and check if polluted properties exist
      const testObj = {};
      return testObj.polluted === true || 
             testObj.toString === 'polluted' || 
             this.hasPollutedProperties(testObj);
    } catch (e) {
      return false;
    }
  }

  hasPollutedProperties(obj) {
    // Check for unexpected properties in object prototype
    for (const key in obj) {
      if (!obj.hasOwnProperty(key) && key === 'polluted') {
        return true;
      }
    }
    return false;
  }

  getPollutedProperty(data) {
    const testObj = {};
    if (testObj.polluted === true) return 'polluted';
    if (testObj.toString === 'polluted') return 'toString';
    
    for (const key in testObj) {
      if (!testObj.hasOwnProperty(key)) {
        return key;
      }
    }
    return 'unknown';
  }
}

// Usage example
(async () => {
  const scanner = new PrototypePollutionScanner('http://example.com/api', {
    verbose: true,
    method: 'POST'
  });
  
  await scanner.scan();
})();

File: src/scanner.py
import requests
import json
import time

class PrototypePollutionScanner:
    def __init__(self, target_url, options=None):
        self.target_url = target_url
        self.options = options or {}
        self.options.setdefault('method', 'POST')
        self.options.setdefault('headers', {'Content-Type': 'application/json'})
        self.options.setdefault('timeout', 5)
        self.options.setdefault('verbose', False)
        self.vulnerable = False
        self.test_payloads = self.generate_test_payloads()

    def generate_test_payloads(self):
        return [
            {'__proto__': {'polluted': True}},
            {'constructor': {'prototype': {'polluted': True}}},
            {'constructor': {'prototype': {'toString': 'polluted'}}},
            {'prototype': {'polluted': True}}
        ]

    async def scan(self):
        print(f"[+] Starting prototype pollution scan for {self.target_url}")
        
        for payload in self.test_payloads:
            try:
                if self.options.get('verbose'):
                    print(f"[.] Testing payload: {json.dumps(payload)}")

                response = requests.request(
                    method=self.options['method'],
                    url=self.target_url,
                    json=payload,
                    headers=self.options['headers'],
                    timeout=self.options['timeout']
                )

                # Check if pollution was successful
                check_response = requests.get(self.target_url)
                
                if self.check_for_pollution(check_response.json()):
                    self.vulnerable = True
                    print(f"[!] Prototype pollution vulnerability found with payload: {json.dumps(payload)}")
                    print(f"[!] Affected property: {self.get_polluted_property(check_response.json())}")
                    return True

            except Exception as error:
                if self.options.get('verbose'):
                    print(f"[!] Error testing payload {json.dumps(payload)}: {str(error)}")

        if not self.vulnerable:
            print("[+] No prototype pollution vulnerabilities detected")
        return False

    def check_for_pollution(self, data):
        try:
            test_obj = {}
            return (hasattr(test_obj, 'polluted') and test_obj.polluted is True) or \
                   (hasattr(test_obj, 'toString') and test_obj.toString == 'polluted') or \
                   self.has_polluted_properties(test_obj)
        except:
            return False

    def has_polluted_properties(self, obj):
        for key in obj:
            if not obj.hasOwnProperty(key) and key == 'polluted':
                return True
        return False

    def get_polluted_property(self, data):
        test_obj = {}
        if hasattr(test_obj, 'polluted') and test_obj.polluted is True:
            return 'polluted'
        if hasattr(test_obj, 'toString') and test_obj.toString == 'polluted':
            return 'toString'
        
        for key in test_obj:
            if not test_obj.hasOwnProperty(key):
                return key
        return 'unknown'

# Usage example
if __name__ == "__main__":
    scanner = PrototypePollutionScanner('http://example.com/api', {
        'verbose': True,
        'method': 'POST'
    })
    
    scanner.scan()
